Sources/BuildSystemIntegration/SwiftPMBuildSystem.swift

```diff
@@ -394,45 +396,26 @@ extension SwiftPMBuildSystem {
     /// Make sure to execute any throwing statements before setting any
     /// properties because otherwise we might end up in an inconsistent state
     /// with only some properties modified.
-    self.modulesGraph = modulesGraph
-
-    self.targets = Dictionary(
-      try buildDescription.allTargetsInTopologicalOrder(in: modulesGraph).enumerated().map { (index, target) in
-        return (key: ConfiguredTarget(target), value: (index, target))
-      },
-      uniquingKeysWith: { first, second in
-        logger.fault("Found two targets with the same name \(first.buildTarget.name)")
-        return second
-      }
-    )
 
-    self.fileToTargets = [DocumentURI: [SwiftBuildTarget]](
-      modulesGraph.allModules.flatMap { target in
-        return target.sources.paths.compactMap { (filePath) -> (key: DocumentURI, value: [SwiftBuildTarget])? in
-          guard let buildTarget = buildDescription.getBuildTarget(for: target, in: modulesGraph) else {
-            return nil
-          }
-          return (key: DocumentURI(filePath.asURL), value: [buildTarget])
-        }
-      },
-      uniquingKeysWith: { $0 + $1 }
-    )
+    self.targets = [:]
+    self.fileToTargets = [:]
+    buildDescription.traverseModules { buildTarget, parent, depth in
+      let configuredTarget = ConfiguredTarget(buildTarget)
+      var depth = depth
+      if let existingDepth = targets[configuredTarget]?.depth {
+        depth = max(existingDepth, depth)
+      }
+      targets[configuredTarget] = (buildTarget, depth)
 
-    self.sourceDirToTargets = [DocumentURI: [SwiftBuildTarget]](
-      modulesGraph.allModules.compactMap { (target) -> (DocumentURI, [SwiftBuildTarget])? in
-        guard let buildTarget = buildDescription.getBuildTarget(for: target, in: modulesGraph) else {
-          return nil
-        }
-        return (key: DocumentURI(target.sources.root.asURL), value: [buildTarget])
-      },
-      uniquingKeysWith: { $0 + $1 }
-    )
+      for source in buildTarget.sources + buildTarget.headers {
```

Comment:
I think if we have already saw a particular target we can skip doing this since all of the references would produce the same source and headers regardless of what depends on it...

Reply:
Oh yes. Thanks for noticing.